### PR TITLE
Enhance IdSet virtual table performance with optimized point lookups and IN filtering

### LINKED ISSUE #9233: IdSet virtual table: add optimized query test cases (user request)
## Context

This issue tracks a user request raised in PR #9229 — _Enhance IdSet virtual table performance with optimized point lookups and IN filtering_.

The PR improves `IdSet` virtual table performance by switching internal storage to a sorted vector, enabling O(log n) point lookups and efficient `IN` filtering. During review, a request was made to provide additional **optimized query cases** so consumers can understand and validate the performance characteristics of the new implementation.

## Request

Add or document optimized test / example cases that cover:

- [x] Point lookup with large input sets (stress test O(log n) behavior)
- [x] `IN` filtering with varying selectivity (few hits vs many hits)
- [x] Sorted/deduplicated output with random and reverse-ordered input
- [x] Edge cases: empty set, single element, all duplicates
- [x] Comparison/benchmark against the previous full-scan behavior to illustrate the improvement

## Related

- PR: https://github.com/iTwin/itwinjs-core/pull/9229
- imodel-native PR: https://github.com/iTwin/imodel-native/pull/1392

### PR DESCRIPTION
**Pull request by khanaffan**
imodel-native: https://github.com/iTwin/imodel-native/pull/1392

## Summary

Improves the performance of the `IdSet` virtual table by switching its internal storage to a **sorted vector**. This enables:

- **O(log n) point lookups** for `id = ?` queries via binary search instead of full scans
- **Efficient `IN` filtering** using a single-pass filter call rather than per-value scans
- **Sorted, deduplicated output** regardless of input order

## Changes

- `IdSetVTTests.ecsql.md`: Added test cases covering point lookups (hit and miss), `IN` optimization (hit and miss), and sorted/deduplicated output
- `docs/changehistory/NextVersion.md`: Documented the improvement under `@itwin/core-backend`

## Examples

```sql
-- Point lookup — binary search, O(log n)
SELECT id FROM IdSet('[1,2,3,4,5]') WHERE id = 3
-- returns: 0x3

-- IN filter — single Filter call for all values
SELECT id FROM IdSet('[1,2,3,4,5,6,7,8,9,10]') WHERE id IN (3, 5, 7)
-- returns: 0x3, 0x5, 0x7

-- Sorted deduplication — unsorted input with duplicates
SELECT id FROM IdSet('[50,10,30,20,40,10]')
-- returns: 0xa, 0x14, 0x1e, 0x28, 0x32
```

## Validation

- Targeted verification: New ECSQL test cases added for point lookup, IN filtering, and deduplication behavior
- Known baseline issues: None

### CODE REVIEW

**Review by rschili — APPROVED** [review]

File: core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md [hunk 1]
```diff
@@ -731,3 +731,87 @@ SELECT id FROM IdSet
...
+# Point lookup: id = literal returns single matching row
+
+- dataset: AllProperties.bim
+
+```sql
+SELECT id FROM IdSet('[1,2,3,4,5]') WHERE id = 3
+```
+
+| className | accessString | generated | index | jsonName | name | extendedType | typeName | type | originPropertyName |
+| --------- | ------------ | --------- | ----- | -------- | ---- | ------------ | -------- | ---- | ------------------ |
```
Comment (hl662) [inline core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md:744]: The new point-lookup tests exercise `xFilter` directly, but the production pattern could be a join (`WHERE id = ECInstanceId`), which also goes through `xBestIndex` cost estimation. Might be worth adding at least one join-form test to make sure the optimizer still picks IdSet as the inner loop correctly? Something like:
```sql
SELECT e.i FROM aps.TestElement e, IdSet('[21,24,25]') WHERE e.ECInstanceId = id AND id = 21
```
This mirrors the existing test at line 192 and would catch any regression in join-plan cost estimation. 🤖

Comment (hl662) [inline core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md:744]: All five new tests use inline JSON literals (`IdSet('[1,2,3,4,5]')`), but the existing suite is almost entirely `bindIdSet` + `IdSet(?)`. There's even an explicit test (line 501) documenting that the two paths behave differently in QueryReader mode. Could you add at least one `bindIdSet` counterpart for the point-lookup case to make sure the sorted-vector refactor applies to the bound path too?
```
- dataset: AllProperties.bim
- bindIdSet 1, [1, 2, 3, 4, 5]

```sql
SELECT id FROM IdSet(?) WHERE id = 3
```
Expected: `0x3` 🤖

File: docs/changehistory/NextVersion.md [hunk 2]
```diff
@@ -53,6 +54,20 @@ This is intentional — it avoids a class of bugs where `removeListener()` silen
 
 ## @itwin/core-backend
 
+### IdSet virtual table performance improvements
+
+The `IdSet` virtual table now uses a sorted vector internally, enabling O(log n) point lookups (`id = ?`) and efficient single-pass `IN` filtering instead of full scans.
+
+```sql
+-- Point lookup — binary search, O(log n)
+SELECT i FROM aps.TestElement, IdSet('[1,2,3,4,5]') WHERE id = ECInstanceId AND id = 3
```
Comment (hl662) [inline docs/changehistory/NextVersion.md]: The SQL examples reference `aps.TestElement` and the `i` column, which are from the internal AllProperties test schema — a user copying this into their own project would get a prepare error. Could these either be simplified to match the actual new tests (just `SELECT id FROM IdSet(...)`) or use a schema everyone has, like `bis.Element` with `ECInstanceId`? e.g.:
```sql
-- Point lookup — binary search, O(log n)
SELECT e.ECInstanceId FROM bis.Element e, IdSet('[1,2,3,4,5]') v
  WHERE e.ECInstanceId = v.id AND v.id = 3
```
🤖

File: docs/changehistory/NextVersion.md [hunk 3]
```diff
@@ -53,6 +54,20 @@ This is intentional — it avoids a class of bugs where `removeListener()` silen
...
+
+-- IN filter — single Filter call for all values
+SELECT i FROM aps.TestElement, IdSet('[1,2,3,4,5,6,7,8,9,10]') WHERE id = ECInstanceId AND id IN (3, 5, 7)
+```
+
+Output is always returned sorted and deduplicated, regardless of input order.
```
Comment (hl662) [inline docs/changehistory/NextVersion.md]: Might be worth clarifying that "sorted" means ascending by ID value, and flagging this as a behavioral change? Something like: *"Output is always returned in ascending ID order and deduplicated, regardless of input order. This is a behavioral change — previously, output order was unspecified."* 🤖

**Review by hl662 — COMMENTED** [review]
Review of test cases and documentation for the IdSet vtab perf improvements. A few things worth discussing before merge.

**Comment by hl662** [comment]
Curious, do we have a test case for when we pass in an empty `[]`? And what typically happens there 🤖

File: core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md (hunk 1 above)
Comment (khanaffan) [reply core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md:744]: honestly there was no point adding these test as my major motivation was to update NextVersion.md. Few of following test make sense but we have native test that better test this. Its not a syntax change. It only effect execution plan.

Comment (hl662) [reply core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md:744]: would these tests be valuable as examples to be shown on the docs site?

Comment (khanaffan) [reply core/backend/src/test/ecsql/queries/IdSetVTTests.ecsql.md:744]: I am not sure. Again it's an internal optimization. The query worked before but produce suboptimal plan.

**Review by hl662 — APPROVED** [review]
imodel-native has the tests so I'm not worried about lack of it. However I am starting to think about how we might want to add more good ecsql examples in our learning docs. Especially now that we are seeing traction with our docs mcp

**Comment by DanRod1999** [comment]
This PR is linked to native, but an issue hit in the pipeline so the new version of the addon was not pushed. However, this branch now is in line with master, and master branch is on imodel-native 5.9.14 which is the correct addon version. This should be good to merge

**Review by aruniverse — APPROVED** [review]

**Comment by DanRod1999** [comment]
I think in the future we don't need to make PR's dependent on each other just for md file changes like this, so instead of linking the PRs, we can just say "this documentation is in response to changes from this imodel-native PR" or something like that. This way we don't need to waste time on builds, or have reviews expect to see a new addon merge into this PR.
